pkg/istio-agent/xds_proxy.go

```diff
@@ -210,7 +215,7 @@ func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDisc
 				proxyLog.Warnf("upstream terminated with unexpected error %v", err)
 			}
 			_ = upstream.CloseSend()
-			return err
+			goto RecreateUpstream
```

Comment:
Can we add some tests? I am not sure this will work properly. When `upstreamError` is hit, this may be from the server closing. But we will try to reuse the connection, which will fail.

We either need more robust handling of connection errors, or we should just recreate the connection like current code but make sure we close the old one

Reply:
I think the grpc client will always reconnect to the server when any error occur.

https://github.com/grpc/grpc-go/blob/3c400e7fcc8719ba3d133563f9c9d126f03a80c0/clientconn.go#L825

Reply:
>When upstreamError is hit, this may be from the server closing. But we will try to reuse the connection, which will fail.

We may receive some error at this point and we will recreate a new stream. 

>Can we add some tests? 

I am not sure what kind of test you mean. I think this requires long-running test.